src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+-- split as a valid path, i.e split at a path separator only.
+splitPathAt :: Int -> ByteString -> (ShortByteString, ShortByteString)
+splitPathAt n fp
+    | S.length fp <= n = (SS.empty, toShort fp)
+    | otherwise =
+        let sfp = S8.splitWith isPathSeparator fp
+            sepWith p (tlen, prefix', suffix') =
+                case S.length p + 1 + tlen of
+                    tlen'
+                        | tlen' <= n -> (tlen', prefix', p : suffix')
+                    tlen' -> (tlen', p : prefix', suffix')
+            (_, prefix, suffix) = foldr' sepWith (0, [], []) sfp
+            toShortPath = toShort . S8.intercalate pathSeparatorS
+        in (toShortPath prefix, toShortPath suffix)
+
+packHeader :: MonadThrow m => Header -> m S.ByteString
+packHeader header = do
+    (left, right) <- packHeaderNoChecksum header
+    let sumsl :: SL.ByteString -> Int
+        sumsl = SL.foldl' (\ !acc !v -> acc + fromIntegral v) 0
+    encChecksum <- encodeOctal 7 $ sumsl left + 32 * 8 + sumsl right
+    return $
+        SL.toStrict $
+        toLazyByteString $ lazyByteString left <> encChecksum <> word8 0 <> lazyByteString right
+
+
+packHeaderNoChecksum :: MonadThrow m => Header -> m (SL.ByteString, SL.ByteString)
+packHeaderNoChecksum Header {..} = do
+    let CTime headerTime' = headerTime
+    hNameSuffix <- encodeShort 100 headerFileNameSuffix
+    hFileMode <- encodeOctal 7 headerFileMode
+    hOwnerId <- encodeOctal 7 headerOwnerId
+    hGroupId <- encodeOctal 7 headerGroupId
+    hPayloadSize <- encodeOctal 11 headerPayloadSize
+    hTime <- encodeOctal 11 headerTime'
+    hLinkName <- encodeShort 100 headerLinkName
+    hMagicVersion <- encodeShort 8 headerMagicVersion
+    hOwnerName <- encodeShort 32 headerOwnerName
+    hGroupName <- encodeShort 32 headerGroupName
+    hDevMajor <- encodeDevice headerDeviceMajor
+    hDevMinor <- encodeDevice headerDeviceMinor
+    hNamePrefix <- encodeShort 155 headerFileNamePrefix
+    return
+        ( toLazyByteString $
+          hNameSuffix <>
+          hFileMode <> word8 0 <>
+          hOwnerId <> word8 0 <>
+          hGroupId <> word8 0 <>
+          hPayloadSize <> word8 0 <>
+          hTime <> word8 0
+        , toLazyByteString $
+          word8 headerLinkIndicator <>
+          hLinkName <>
+          hMagicVersion <>
+          hOwnerName <>
+          hGroupName <>
+          hDevMajor <> word8 0 <>
+          hDevMinor <> word8 0 <>
+          hNamePrefix <>
+          byteString (S.replicate 12 0)
+        )
+  where
+    encodeDevice 0     = return $ byteString $ S.replicate 7 0
+    encodeDevice devid = encodeOctal 7 devid
+
+
+-- | Encode a `ShortByteString` with an exact length, NUL terminating if it is
+-- shorter, but throwing `TarCreationError` if it is longer.
+encodeShort :: MonadThrow m => Int -> ShortByteString -> m Builder
+encodeShort !len !sbs
+    | lenShort <= len = return $ byteString $ fst $ S.unfoldrN len maybeShort 0
```

Comment:
Any particular reason not to use `shortByteString` here? https://www.stackage.org/haddock/lts-10.3/bytestring-0.10.8.2/Data-ByteString-Builder.html#v:shortByteString

Reply:
Now that I look at this a month after I wrote it, there is a much easier way to write this function, just as you pointed it out. Good catch. New version:
```
    | lenShort <= len = return $ shortByteString sbs <> byteString (S.replicate (len - lenShort) 0)
```

Reply:
I got way too intrigued by what the fastest way would be to create a certain amount of null padding and wrote this benchmark:

https://gist.github.com/snoyberg/a8aa56183a1bf50238482b0bd3675fb0

Surprisingly to me, it seems like the most consistently well-performing approach is to use `S.replicate` as you've done here. I would have hoped that builders would have some way of doing this more efficiently, but I can't discover what it would be.